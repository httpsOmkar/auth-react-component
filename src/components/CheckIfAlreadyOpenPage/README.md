````js
<>
    <CheckIfAlreadyOpenPage errorChildren={ <p>Error</p> } children={ <p>I first</p> }>
    </CheckIfAlreadyOpenPage>
    <CheckIfAlreadyOpenPage errorChildren={ <p>Error</p> } children={ <p>I first</p> }>
    </CheckIfAlreadyOpenPage>
</>
````
