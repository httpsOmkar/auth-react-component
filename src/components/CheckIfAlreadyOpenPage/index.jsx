import React from 'react';
import { element } from 'prop-types';

/**
 * This component for reopening the same page in two or more tabs.
 * To avoid the token errors
 */
export default class CheckIfAlreadyOpenPage extends React.Component {
  static propTypes = {
    /**
     * Render children when the page is already open
     */
    errorChildren: element.isRequired,

    /**
     * Children when there is no error
     */
    children: element.isRequired,
  };

  static defaultProps = {
    errorChildren: null,
    children: null,
  };

  state = {
    ready: false,
    alreadyOpen: false,
  };

  componentDidMount() {
    if (localStorage && window) {
      localStorage.setItem('page_opened', Date.now().toString());
      const onLocalStorageEvent = event => {
        if (event.key === 'page_opened') {
          // Emit that you're already available.
          return localStorage.setItem('already_open', Date.now().toString());
        }

        if (event.key === 'already_open') {
          return this.setState({ alreadyOpen: true });
        }

        this.setState({ ready: true });
      };

      window.addEventListener('storage', onLocalStorageEvent, false);
    }
  }

  render() {
    const { alreadyOpen, ready } = this.state;
    const { errorChildren, children } = this.props;

    if (ready) {
      return alreadyOpen ? errorChildren : children;
    }

    return <></>;
  }
}
