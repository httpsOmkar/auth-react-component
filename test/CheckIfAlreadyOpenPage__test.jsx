import React from 'react';
import { mount } from 'enzyme';
import CheckIfAlreadyOpenPage from '../src/components/CheckIfAlreadyOpenPage';

describe('<CheckIfAlreadyOpen />', () => {
  it('should render', () => {
    const component = mount(
      <CheckIfAlreadyOpenPage errorChildren={<p>Error</p>}>
        <p>Test</p>
      </CheckIfAlreadyOpenPage>
    );

    expect(component).toMatchSnapshot();
  });
});
