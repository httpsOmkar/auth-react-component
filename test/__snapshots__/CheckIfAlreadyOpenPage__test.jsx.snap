// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckIfAlreadyOpen /> should render 1`] = `
<CheckIfAlreadyOpenPage
  errorChildren={
    <p>
      Error
    </p>
  }
/>
`;
